psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +976,432 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+            size += primary_->shell(P).nfunction();
+        }
+        max_functions_per_atom = std::max(max_functions_per_atom, size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> LinK: Atom Blocking <==\n\n");
+        for (size_t atom = 0; atom < natom; atom++) {
+            outfile->Printf("  Atom: %3d, Atom Start: %4d, Atom End: %4d\n", atom, shell_endpoints_for_atom[atom],
+                            shell_endpoints_for_atom[atom + 1]);
+            for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = basis_endpoints_for_shell[P];
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // => Store significant atom pairs for parallelization <= //
+    std::vector<std::pair<int, int>> atom_pairs;
+    for (size_t Patom = 0; Patom < natom; Patom++) {
+        for (size_t Qatom = 0; Qatom <= Patom; Qatom++) {
+            bool found = false;
+            for (int P = shell_endpoints_for_atom[Patom]; P < shell_endpoints_for_atom[Patom + 1]; P++) {
+                for (int Q = shell_endpoints_for_atom[Qatom]; Q < shell_endpoints_for_atom[Qatom + 1]; Q++) {
+                    if (ints[0]->shell_pair_significant(P, Q)) {
+                        found = true;
+                        atom_pairs.emplace_back(Patom, Qatom);
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+    }
+
+    // => Store shell pairs that are significant by Schwarz estimate <= //
+    // A comparator used for sorting integral screening values
+    auto screen_compare = [](const std::pair<int, double> &a, 
+                                    const std::pair<int, double> &b) { return a.second > b.second; };
+
+    std::vector<std::vector<int>> significant_bras(nshell);
+    double max_integral = ints[0]->max_integral();
+
+    for (size_t P = 0; P < nshell; P++) {
+        std::vector<std::pair<int, double>> PQ_shell_values;
+        for (size_t Q = 0; Q < nshell; Q++) {
+            double pq_pq = std::sqrt(ints[0]->shell_ceiling2(P, Q, P, Q));
+            double schwarz_value = std::sqrt(pq_pq * max_integral);
+            if (schwarz_value >= cutoff_) {
+                PQ_shell_values.emplace_back(Q, schwarz_value);
+            }
+        }
+        std::sort(PQ_shell_values.begin(), PQ_shell_values.end(), screen_compare);
+
+        for (const auto& value : PQ_shell_values) {
+            significant_bras[P].push_back(value.first);
+        }
+    }
+
+    // => Calculate Shell Ceilings (To find significant bra-ket pairs) <=
```

Comment:
Done. The shell ceilings code is moved to be a subitem of `// => "Pre-ordering and Pre-selection to find significant elements in Puv" <= //`. I was not able to combine the two `P < nshell` loops, since it would interfere with the logic of the `significant_kets` code.